File: examples/js/product-consumer/provider-server.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { AddressInfo } from 'node:net';

import cors from 'cors';
import express from 'express';
import { ApolloServer } from '@apollo/server';
import { expressMiddleware } from '@apollo/server/express4';

const schema = readFileSync(resolve(__dirname, 'schema.graphql'), 'utf8');

const data = {
  categories: [
    { id: 'cat-1', name: 'Bags', slug: 'bags' },
    { id: 'cat-2', name: 'Travel', slug: 'travel' },
  ],
  products: [
    {
      id: 'prod-1',
      sku: 'SKU-TRAIL-001',
      name: 'Trail Backpack',
      description: 'Lightweight hiking pack',
      status: 'ACTIVE',
      categoryId: 'cat-1',
      tags: ['outdoor', 'hiking'],
      variantIds: ['var-1'],
      reviewIds: ['rev-1'],
    },
    {
      id: 'prod-2',
      sku: 'SKU-WEEK-004',
      name: 'Weekender Tote',
      description: 'Carry-on sized tote',
      status: 'DRAFT',
      categoryId: 'cat-2',
      tags: ['travel', 'weekend'],
      variantIds: ['var-2'],
      reviewIds: [],
    },
  ],
  variants: [
    {
      id: 'var-1',
      sku: 'SKU-TRAIL-001',
      name: 'Trail Backpack - Black',
      price: { list: { amount: 129.99, currency: 'USD' }, sale: null },
      inventory: { quantity: 42, updatedAt: '2026-03-08T12:00:00Z' },
    },
    {
      id: 'var-2',
      sku: 'SKU-WEEK-004',
      name: 'Weekender Tote - Sand',
      price: { list: { amount: 159.99, currency: 'USD' }, sale: null },
      inventory: { quantity: 18, updatedAt: '2026-03-08T12:00:00Z' },
    },
  ],
  customers: [
    {
      id: 'cust-1',
      email: 'alex@example.com',
      name: 'Alex',
      defaultShippingAddress: {
        line1: '123 Market St',
        line2: null,
        city: 'San Francisco',
        state: 'CA',
        postalCode: '94103',
        country: 'US',
      },
    },
  ],
  reviews: [
    {
      id: 'rev-1',
      rating: 'FIVE',
      title: 'Perfect for trips',
      body: 'Roomy and comfortable.',
      authorId: 'cust-1',
      createdAt: '2026-03-01T08:00:00Z',
    },
  ],
  orders: [
    {
      id: 'order-1',
      status: 'PLACED',
      customerId: 'cust-1',
      lineItems: [
        { productId: 'prod-1', variantId: 'var-1', quantity: 2 },
        { productId: 'prod-2', variantId: 'var-2', quantity: 1 },
      ],
      total: { amount: 289.97, currency: 'USD' },
      shipments: [
        {
          id: 'ship-1',
          status: 'PENDING',
          trackingNumber: null,
          carrier: null,
          address: {
            line1: '123 Market St',
            line2: null,
            city: 'San Francisco',
            state: 'CA',
            postalCode: '94103',
            country: 'US',
          },
          shippedAt: null,
          deliveredAt: null,
        },
      ],
      placedAt: '2026-03-08T09:00:00Z',
    },
  ],
};

const getProduct = (id: string) => data.products.find((product) => product.id === id);
const getVariant = (id: string) => data.variants.find((variant) => variant.id === id);
const getCategory = (id: string) => data.categories.find((category) => category.id === id);
const getCustomer = (id: string) => data.customers.find((customer) => customer.id === id);
const getOrder = (id: string) => data.orders.find((order) => order.id === id);

export function buildInventoryChangedEvent(variantId: string) {
  const variant = getVariant(variantId);
  if (!variant) {
    throw new Error('Variant not found');
  }
  return {
    subscription: 'InventoryChanged',
    variables: { variantId },
    data: { inventoryChanged: variant.inventory },
  };
}

const resolveNodeType = (obj: { id?: string }) => {
  if (!obj.id) {
    return null;
  }
  if (obj.id.startsWith('prod-')) {
    return 'Product';
  }
  if (obj.id.startsWith('var-')) {
    return 'Variant';
  }
  if (obj.id.startsWith('cat-')) {
    return 'Category';
  }
  if (obj.id.startsWith('rev-')) {
    return 'Review';
  }
  if (obj.id.startsWith('cust-')) {
    return 'Customer';
  }
  if (obj.id.startsWith('ship-')) {
    return 'Shipment';
  }
  if (obj.id.startsWith('order-')) {
    return 'Order';
  }
  return null;
};

const resolveSearchResultType = (obj: { id?: string }) => {
  if (!obj.id) {
    return null;
  }
  if (obj.id.startsWith('prod-')) {
    return 'Product';
  }
  if (obj.id.startsWith('cat-')) {
    return 'Category';
  }
  if (obj.id.startsWith('rev-')) {
    return 'Review';
  }
  return null;
};

const resolvers = {
  Query: {
    product: (_: unknown, args: { id: string }) => {
      if (args.id === '10') {
        return {
          id: '10',
          sku: 'SKU-10',
          name: 'product name',
          description: null,
          status: 'ACTIVE',
          categoryId: 'cat-1',
          tags: [],
          variantIds: [],
          reviewIds: [],
        };
      }
      return getProduct(args.id) ?? null;
    },
    products: (_: unknown, args: { status?: string; search?: string }) => {
      let results = data.products;
      if (args.status) {
        results = results.filter((product) => product.status === args.status);
      }
      if (args.search) {
        const term = args.search.toLowerCase();
        results = results.filter((product) =>
          [product.name, product.description ?? '', product.sku].some((field) =>
            field.toLowerCase().includes(term),
          ),
        );
      }
      return results;
    },
    search: (_: unknown, args: { text: string }) => {
      const term = args.text.toLowerCase();
      const productMatches = data.products.filter((product) =>
        product.name.toLowerCase().includes(term),
      );
      const categoryMatches = data.categories.filter((category) =>
        category.name.toLowerCase().includes(term),
      );
      const reviewMatches = data.reviews.filter((review) =>
        (review.title ?? '').toLowerCase().includes(term),
      );
      return [...productMatches, ...categoryMatches, ...reviewMatches];
    },
    order: (_: unknown, args: { id: string }) =>
      getOrder(args.id) ?? null,
  },
  Mutation: {
    addToCart: (
      _: unknown,
      args: { input: { productId: string; variantId?: string; quantity: number } },
    ) => {
      const product = getProduct(args.input.productId);
      if (!product) {
        throw new Error('Product not found');
      }
      const variantId = args.input.variantId ?? product.variantIds[0];
      if (!variantId) {
        throw new Error('Variant not found');
      }
      const variant = getVariant(variantId);
      if (!variant) {
        throw new Error('Variant not found');
      }
      return {
        product,
        variant,
        quantity: args.input.quantity,
        price: variant.price,
      };
    },
    placeOrder: (
      _: unknown,
      args: {
        input: {
          customerId: string;
          lines: Array<{ productId: string; variantId?: string; quantity: number }>;
          shippingAddress: {
            line1: string;
            line2?: string | null;
            city: string;
            state: string;
            postalCode: string;
            country: string;
          };
        };
      },
    ) => {
      const customer = getCustomer(args.input.customerId);
      if (!customer) {
        throw new Error('Customer not found');
      }

      const orderLines = args.input.lines.map((line) => {
        const product = getProduct(line.productId);
        if (!product) {
          throw new Error('Product not found');
        }
        const variantId = line.variantId ?? product.variantIds[0];
        if (!variantId) {
          throw new Error('Variant not found');
        }
        const variant = getVariant(variantId);
        if (!variant) {
          throw new Error('Variant not found');
        }
        return { productId: product.id, variantId: variant.id, quantity: line.quantity };
      });

      return {
        id: 'order-1',
        status: 'PLACED',
        customerId: customer.id,
        lineItems: orderLines,
        total: { amount: 289.97, currency: 'USD' },
        shipments: [
          {
            id: 'ship-1',
            status: 'PENDING',
            trackingNumber: null,
            carrier: null,
            address: {
              line1: args.input.shippingAddress.line1,
              line2: args.input.shippingAddress.line2 ?? null,
              city: args.input.shippingAddress.city,
              state: args.input.shippingAddress.state,
              postalCode: args.input.shippingAddress.postalCode,
              country: args.input.shippingAddress.country,
            },
            shippedAt: null,
            deliveredAt: null,
          },
        ],
        placedAt: '2026-03-08T09:00:00Z',
      };
    },
    submitReview: (
      _: unknown,
      args: { input: { productId: string; rating: string; title?: string | null; body?: string | null } },
    ) => {
      const product = getProduct(args.input.productId);
      if (!product) {
        throw new Error('Product not found');
      }
      const author = data.customers[0];
      return {
        id: 'rev-new',
        rating: args.input.rating,
        title: args.input.title ?? null,
        body: args.input.body ?? null,
        authorId: author.id,
        createdAt: '2026-03-08T10:00:00Z',
      };
    },
  },
  Product: {
    category: (product: { categoryId: string }) => getCategory(product.categoryId) ?? null,
    variants: (product: { variantIds: string[] }) =>
      data.variants.filter((variant) => product.variantIds.includes(variant.id)),
    reviews: (product: { reviewIds: string[] }) =>
      data.reviews.filter((review) => product.reviewIds.includes(review.id)),
  },
  Variant: {
    price: (variant: { price: unknown }) => variant.price,
    inventory: (variant: { inventory: unknown; id: string }) =>
      buildInventoryChangedEvent(variant.id).data.inventoryChanged,
  },
  Review: {
    author: (review: { authorId: string }) => {
      const author = getCustomer(review.authorId);
      if (!author) {
        throw new Error('Customer not found');
      }
      return author;
    },
  },
  Customer: {
    defaultShippingAddress: (customer: { defaultShippingAddress: unknown }) =>
      customer.defaultShippingAddress,
    addresses: (customer: { defaultShippingAddress: unknown }) =>
      customer.defaultShippingAddress ? [customer.defaultShippingAddress] : [],
  },
  Order: {
    customer: (order: { customerId: string }) => {
      const customer = getCustomer(order.customerId);
      if (!customer) {
        throw new Error('Customer not found');
      }
      return customer;
    },
    lines: (order: { lineItems: Array<{ productId: string; variantId: string; quantity: number }> }) =>
      order.lineItems.map((line) => {
        const product = getProduct(line.productId);
        if (!product) {
          throw new Error('Product not found');
        }
        const variant = getVariant(line.variantId);
        if (!variant) {
          throw new Error('Variant not found');
        }
        return {
          product,
          variant,
          quantity: line.quantity,
          price: variant.price,
        };
      }),
    total: (order: { total: unknown }) => order.total,
    shipments: (order: { shipments: unknown }) => order.shipments,
  },
  SearchResult: {
    __resolveType: resolveSearchResultType,
  },
  Node: {
    __resolveType: resolveNodeType,
  },
};

export async function startProviderServer() {
  const server = new ApolloServer({ typeDefs: schema, resolvers });
  await server.start();

  const app = express();
  app.use(
    '/graphql',
    cors<cors.CorsRequest>(),
    express.json({ type: ['application/json', 'application/graphql'] }),
    expressMiddleware(server),
  );

  const httpServer = app.listen(0);
  const { port } = httpServer.address() as AddressInfo;

  return {
    url: `http://127.0.0.1:${port}`,
    close: async () => {
      await server.stop();
      await new Promise<void>((resolve) => httpServer.close(() => resolve()));
    },
  };
}

File: examples/go/product-consumer/provider_server.go
package productconsumer

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
)

type graphqlRequest struct {
	Query         string         `json:"query"`
	Variables     map[string]any `json:"variables"`
	OperationName string         `json:"operationName"`
}

type providerServer struct {
	url    string
	server *http.Server
}

type inventoryEvent struct {
	Subscription string         `json:"subscription"`
	Variables    map[string]any `json:"variables,omitempty"`
	Data         map[string]any `json:"data"`
}

func buildInventoryChangedEvent(variantID string) (inventoryEvent, error) {
	variant, ok := providerData.variants[variantID]
	if !ok {
		return inventoryEvent{}, errors.New("variant not found")
	}

	return inventoryEvent{
		Subscription: "InventoryChanged",
		Variables:    map[string]any{"variantId": variantID},
		Data: map[string]any{
			"inventoryChanged": map[string]any{
				"quantity":  variant.Inventory.Quantity,
				"updatedAt": variant.Inventory.UpdatedAt,
			},
		},
	}, nil
}

func startProviderServer() (*providerServer, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return nil, err
	}

	mux := http.NewServeMux()
	mux.HandleFunc("/graphql", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			w.WriteHeader(http.StatusMethodNotAllowed)
			return
		}

		var payload graphqlRequest
		if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
			w.WriteHeader(http.StatusBadRequest)
			_, _ = w.Write([]byte(`{"error":"invalid JSON"}`))
			return
		}

		response, err := buildResponse(payload)
		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			_, _ = w.Write([]byte(fmt.Sprintf(`{"error":%q}`, err.Error())))
			return
		}

		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(response)
	})

	server := &http.Server{Handler: mux}
	go func() {
		_ = server.Serve(listener)
	}()

	return &providerServer{
		url:    fmt.Sprintf("http://%s", listener.Addr().String()),
		server: server,
	}, nil
}

func (s *providerServer) close(ctx context.Context) error {
	return s.server.Shutdown(ctx)
}

type providerInventory struct {
	Quantity  int    `json:"quantity"`
	UpdatedAt string `json:"updatedAt"`
}

type providerVariant struct {
	ID        string            `json:"id"`
	SKU       string            `json:"sku"`
	Price     map[string]any    `json:"price"`
	Inventory providerInventory `json:"inventory"`
}

type providerProduct struct {
	ID       string
	SKU      string
	Name     string
	Status   string
	Category map[string]any
	Variants []providerVariant
	Reviews  []map[string]any
}

type providerState struct {
	products map[string]providerProduct
	variants map[string]providerVariant
}

var providerData = providerState{
	products: map[string]providerProduct{
		"prod-1": {
			ID:     "prod-1",
			SKU:    "SKU-TRAIL-001",
			Name:   "Trail Backpack",
			Status: "ACTIVE",
			Category: map[string]any{
				"id":   "cat-1",
				"name": "Bags",
			},
			Variants: []providerVariant{
				{
					ID:  "var-1",
					SKU: "SKU-TRAIL-001",
					Price: map[string]any{
						"list": map[string]any{
							"amount":   129.99,
							"currency": "USD",
						},
					},
					Inventory: providerInventory{
						Quantity:  42,
						UpdatedAt: "2026-03-08T12:00:00Z",
					},
				},
			},
			Reviews: []map[string]any{
				{
					"id":     "rev-1",
					"rating": "FIVE",
					"author": map[string]any{
						"id":   "cust-1",
						"name": "Alex",
					},
				},
			},
		},
		"prod-2": {
			ID:     "prod-2",
			SKU:    "SKU-WEEK-004",
			Name:   "Weekender Tote",
			Status: "DRAFT",
		},
	},
	variants: map[string]providerVariant{
		"var-1": {
			ID:  "var-1",
			SKU: "SKU-TRAIL-001",
			Price: map[string]any{
				"list": map[string]any{
					"amount":   129.99,
					"currency": "USD",
				},
			},
			Inventory: providerInventory{
				Quantity:  42,
				UpdatedAt: "2026-03-08T12:00:00Z",
			},
		},
		"var-2": {
			ID:  "var-2",
			SKU: "SKU-WEEK-004",
			Price: map[string]any{
				"list": map[string]any{
					"amount":   159.99,
					"currency": "USD",
				},
			},
			Inventory: providerInventory{
				Quantity:  18,
				UpdatedAt: "2026-03-08T12:00:00Z",
			},
		},
	},
}

func buildResponse(request graphqlRequest) (map[string]any, error) {
	switch request.OperationName {
	case "GetProduct":
		return map[string]any{
			"data": map[string]any{
				"product": map[string]any{
					"id":     "10",
					"name":   "product name",
					"status": "ACTIVE",
				},
			},
		}, nil
	case "ProductsByStatus":
		product := providerData.products["prod-1"]
		return map[string]any{
			"data": map[string]any{
				"products": []map[string]any{
					{
						"id":       product.ID,
						"name":     product.Name,
						"status":   product.Status,
						"category": product.Category,
						"variants": []map[string]any{
							{
								"id":    product.Variants[0].ID,
								"sku":   product.Variants[0].SKU,
								"price": product.Variants[0].Price,
								"inventory": map[string]any{
									"quantity":  product.Variants[0].Inventory.Quantity,
									"updatedAt": product.Variants[0].Inventory.UpdatedAt,
								},
							},
						},
						"reviews": product.Reviews,
					},
				},
			},
		}, nil
	case "PlaceOrder":
		return map[string]any{
			"data": map[string]any{
				"placeOrder": map[string]any{
					"id":     "order-1",
					"status": "PLACED",
					"total": map[string]any{
						"amount":   289.97,
						"currency": "USD",
					},
					"lines": []map[string]any{
						{
							"quantity": 2,
							"product": map[string]any{
								"id":   "prod-1",
								"name": "Trail Backpack",
							},
							"variant": map[string]any{
								"id":  "var-1",
								"sku": "SKU-TRAIL-001",
							},
						},
						{
							"quantity": 1,
							"product": map[string]any{
								"id":   "prod-2",
								"name": "Weekender Tote",
							},
							"variant": map[string]any{
								"id":  "var-2",
								"sku": "SKU-WEEK-004",
							},
						},
					},
					"shipments": []map[string]any{
						{
							"id":     "ship-1",
							"status": "PENDING",
							"address": map[string]any{
								"city":    "San Francisco",
								"country": "US",
							},
						},
					},
				},
			},
		}, nil
	default:
		return nil, fmt.Errorf("unsupported operation %q", request.OperationName)
	}
}
